cloud-controller-manager/do/kubernetes_lb_validator.go

```diff
@@ -0,0 +1,152 @@
+	Log       logr.Logger
+	GClient   *godo.Client
+	ClusterID string
+	Region    string
+}
+
+// DOKSLBServiceValidator ...
+func (v *KubernetesLBServiceValidator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	svc := &v1.Service{}
+
+	v.Log.V(6).Info("decoding received request")
+	err := v.decoder.Decode(req, svc)
+	if err != nil {
+		v.Log.Error(err, "failed to decode request")
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("failed to decode request: %v", err))
+	}
+
+	v.Log.V(6).Info("checking received request")
+	if svc.Spec.Type != v1.ServiceTypeLoadBalancer || svc.DeletionTimestamp != nil {
+		return admission.Allowed("ignoring the service which is either not a load balancer or is being deleted")
+	}
+
+	// TODO: document the requirement for a droplet-tag annotation for DIY users that need to pass in an identifier to droplets they wish to have behind their lb
+	tag := svc.Annotations["kubernetes.digitalocean.com/droplet-tag"]
+
+	if tag == "" {
+		v.Log.V(6).Info("no tag annotation provided, creating tag using cluster id")
+		tag = fmt.Sprintf("k8s: %v", v.ClusterID)
+	}
+
+	// TODO: these forwarding rules are a placeholder. Further development is required to extract the values from the svc object
+	forwardingRules := []godo.ForwardingRule{
+		{
+			EntryProtocol:  "http",
+			EntryPort:      80,
+			TargetProtocol: "http",
+			TargetPort:     80,
+			CertificateID:  "",
+			TlsPassthrough: false,
+		}}
+
+	lbRequest := v.buildRequest(svc.Name, v.Region, tag, forwardingRules)
+
+	if req.OldObject.Object == nil {
+		if len(svc.Status.LoadBalancer.Ingress) == 0 || svc.Status.LoadBalancer.Ingress[0].IP == "" {
+			err = v.validateCreate(ctx, lbRequest)
+			if err != nil {
+				v.Log.Error(err, "failed to validate lb creation")
+				return admission.Denied(fmt.Sprintf("failed to validate lb creation: %v", err))
+			}
+			return admission.Allowed("valid lb create request")
+		}
+	}
+
+	err = v.decoder.DecodeRaw(req.OldObject, svc)
+	if err != nil {
+		v.Log.Error(err, "failed to decode existing object")
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("failed to decode request: %v", err))
+	}
+
+	currentLBID := svc.Annotations["kubernetes.digitalocean.com/load-balancer-id"]
+	if currentLBID == "" {
+		v.Log.Error(fmt.Errorf("missing loadbalancer id in annotation"), "missing kubernetes.digitalocean.com/load-balancer-id annotation")
+		err = v.validateCreate(ctx, lbRequest)
+		if err != nil {
+			v.Log.Error(err, "failed to validate lb creation")
+			return admission.Denied(fmt.Sprintf("failed to validate lb creation: %v", err))
+		}
+		return admission.Allowed("can not validate lb without annotation: kubernetes.digitalocean.com/load-balancer-id annotation")
+	}
+
+	err = v.validateUpdate(ctx, currentLBID, lbRequest)
+	if err != nil {
+		v.Log.Error(err, "failed to update load balancer")
+		return admission.Denied(fmt.Sprintf("failed to validate load balancer update: %v", err))
```

Comment:
Curious: does the validation API / response semantics allow distinguishing between validation rejections and other errors (e.g., network connectivity problems) that we could/should use to more clearly indicate to the caller why the webhook request failed?

Reply:
I've added a method to check the error status code then use the return to determine whether the admission response should be an error or denial